raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -233,6 +235,68 @@ public void appendLeaderChangeMessage(LeaderChangeMessage leaderChangeMessage, l
         }
     }
 
+
+    public void appendSnapshotHeaderMessage(MetadataSnapshotHeaderRecord snapshotHeaderRecord) {
+        appendLock.lock();
+        try {
+            // Ideally there should be nothing in the batch here.
+            // TODO verify this
+            long currentTimeMs = time.milliseconds();
+            ByteBuffer buffer = memoryPool.tryAllocate(256);
+            if (buffer != null) {
+                MemoryRecords data = MemoryRecords.withSnapshotHeaderRecord(
+                    this.nextOffset,
+                    currentTimeMs,
+                    this.epoch,
+                    buffer,
+                    snapshotHeaderRecord
+                );
+                completed.add(new CompletedBatch<>(
+                    nextOffset,
+                    1,
+                    data,
+                    memoryPool,
+                    buffer
+                ));
+                nextOffset += 1;
+            } else {
+                throw new IllegalStateException("Could not allocate buffer for the metadata snapshot header record.");
+            }
+        } finally {
+            appendLock.unlock();
+        }
+    }
+
+    public void appendSnapshotFooterMessage(MetadataSnapshotFooterRecord snapshotFooterRecord) {
+        appendLock.lock();
+        try {
+            forceDrain();
+            long currentTimeMs = time.milliseconds();
+            ByteBuffer buffer = memoryPool.tryAllocate(256);
+            if (buffer != null) {
+                MemoryRecords data = MemoryRecords.withSnapshotFooterRecord(
+                    this.nextOffset,
+                    currentTimeMs,
+                    this.epoch,
+                    buffer,
+                    snapshotFooterRecord
+                );
+                completed.add(new CompletedBatch<>(
+                    nextOffset,
+                    1,
+                    data,
+                    memoryPool,
+                    buffer
+                ));
+                nextOffset += 1;
+            } else {
+                throw new IllegalStateException("Could not allocate buffer for the metadata snapshot footer record.");
+            }
+        } finally {
+            appendLock.unlock();
+        }
+    }
```

Comment:
There is code duplication between these 3 methods. Let's figure out a way to remove this duplicate code.

Reply:
I agree. Let me fix that.